bpo-32176: Set CO_NOFREE in the code object constructor

ncoghlan opened the pull request:
Previously, CO_NOFREE was set in the compiler, which meant
it could end up being set incorrectly when code objects
were created directly. Setting it in the constructor based
on freevars and cellvars ensures it is always accurate,
regardless of how the code object is defined.


<!-- issue-number: bpo-32176 -->
https://bugs.python.org/issue32176
<!-- /issue-number -->

ncoghlan committed:
bpo-32176: Set CO_NOFREE in the code object constructor

Previously, CO_NOFREE was set in the compiler, which meant
it could end up being set incorrectly when code objects
were created directly. Setting it in the constructor based
on freevars and cellvars ensures it is always accurate,
regardless of how the code object is defined.

ncoghlan committed:
Remove debugging line

serhiy-storchaka commented on Lib/test/test_code.py:
```diff
@@ -141,6 +147,49 @@ def test_newempty(self):
+
+        # Ensure the implicit super() call actually works
+        obj = List([1,2,3])
+        self.assertEqual(obj[0], 1)
```
Does it print to stdout? stdout shouldn't be polluted by tests in non-verbose mode.

How do you know that the injected method actually works instead of the original list method? I suggest to transform the value in `__getitem__`. For example by calling `str()`.

ncoghlan replied:
Done - replaced the print() with the use of an f-string as the getitem result.

serhiy-storchaka commented on Lib/test/test_code.py:
```diff
@@ -141,6 +147,49 @@ def test_newempty(self):
+        self.assertFalse(function.__code__.co_flags & inspect.CO_NOFREE)
+
+        # Ensure the implicit super() call actually works
+        obj = List([1,2,3])
```
Add spaces after commas.

ncoghlan replied:
Done

serhiy-storchaka commented on Lib/test/test_code.py:
```diff
@@ -141,6 +147,49 @@ def test_newempty(self):
+            setattr(cls, name, FunctionType(code, globals(), name, defaults, closure))
+
+        class List(list):
+            def append(self, elem):
```
Are these methods related to the test?

ncoghlan replied:
Just copied from the original reproducer in the bug report - replaced with `pass`.

serhiy-storchaka commented on Lib/test/test_code.py:
```diff
@@ -141,6 +147,49 @@ def test_newempty(self):
+        def new_code(c_or_f):
+            '''A new code object with a __class__ cell added to freevars'''
+            c = c_or_f.__code__ if isinstance(c_or_f, FunctionType) else c_or_f
+            return CodeType(
```
Seems there are no direct tests for CodeType. I think we need to add them (in different issue).

ncoghlan replied:
Yeah, that would probably be a good idea. I did a brief review for other unhandled edge cases in direct calls to the constructor and didn't see any, but I could easily have missed something.

serhiy-storchaka commented on Lib/test/test_code.py:
```diff
@@ -141,6 +147,49 @@ def test_newempty(self):
+
+        def new_code(c_or_f):
+            '''A new code object with a __class__ cell added to freevars'''
+            c = c_or_f.__code__ if isinstance(c_or_f, FunctionType) else c_or_f
```
In this test `new_code()` is called only with the code object.

ncoghlan replied:
I refactored both this and `add_foreign_method` to only handle the cases needed for the test (the original more general versions came from the reported reproducer for the issue)

serhiy-storchaka commented on Lib/test/test_code.py:
```diff
@@ -141,6 +147,49 @@ def test_newempty(self):
+        self.assertIs(class_ref, List)
+
+        # Ensure the code correctly indicates it accesses a free variable
+        self.assertFalse(function.__code__.co_flags & inspect.CO_NOFREE)
```
Add the second argument `hex(function.__code__.co_flags)`.

ncoghlan replied:
Done

ncoghlan committed:
Tidy up the new test case

- use a non-dunder name for the injected function
- remove some code branches not actually needed for the test
- transform the value to ensure the injected method is running
- don't print anything as a side effect of running the test
- print the code flags as hex if the test assertion fails

ncoghlan reviewed (commented):
Thanks for the review! All suggestions accepted, and I spotted a few other opportunities for cleanups while I was editing.
